man/man3/Parent_Node.3
.\"
.\"	HTML Tree
.\"	Parent_Node.3
.\"
.\"	Copyright (C) 1999  Paul J. Lucas
.\"
.\"	This program is free software; you can redistribute it and/or modify
.\"	it under the terms of the GNU General Public License as published by
.\"	the Free Software Foundation; either version 2 of the License, or
.\"	(at your option) any later version.
.\" 
.\"	This program is distributed in the hope that it will be useful,
.\"	but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"	GNU General Public License for more details.
.\" 
.\"	You should have received a copy of the GNU General Public License
.\"	along with this program; if not, write to the Free Software
.\"	Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.\" ---------------------------------------------------------------------------
.\" define code-start macro
.de cS
.sp
.nf
.RS 5
.ft CW
.ta .5i 1i 1.5i 2i 2.5i 3i 3.5i 4i 4.5i 5i 5.5i
..
.\" define code-end macro
.de cE
.ft 1
.RE
.fi
.sp
..
.\" ---------------------------------------------------------------------------
.tr ^
.TH \f3Parent_Node\f1 3 "March 2, 2000" "HTML Tree"
.SH NAME
\f(CWParent_Node\fP \- Node in an HTML tree for HTML elements
that contains child nodes
.SH SYNOPSIS
.ft CW
.nf
class Parent_Node : public virtual HTML_Node {
public:
	typedef std::list< HTML_Node* > child_list;

	Parent_Node( Parent_Node *parent = 0 );
	virtual ~Parent_Node();

	void			add_child( HTML_Node *child );
	child_list^&	children();
	child_list const&	children() const;
	bool			empty() const;
	bool			remove_child( HTML_Node *child );

	// overridden
	virtual void	visit( visitor const&, int depth = 0 );

	// inherited
	Parent_Node*	parent() const;
	void			parent( Parent_Node *new_parent );
};
.fi
.ft 1
.SH DESCRIPTION
\f(CWParent_Node\fP is-an \f(CWHTML_Node\fP
that contains child nodes, i.e., content.
For example, the \f(CWSELECT\f1 element below
is a parent of the newline \f(CWText_Node\fP after the \f(CWSELECT\fP
and all of the \f(CWOPTION\f1 elements:
.cS
<SELECT NAME="menu">		\f2parent child\fP
	<OPTION>Blueberry		\f2child grandchild\fP
	<OPTION>Chocolate		\f2child grandchild\fP
	<OPTION>Raspberry		\f2child grandchild\fP
</SELECT>
.cE
whereas an element such as \f(CWIMG\f1 can have no child nodes.
.SS "Public Interface"
.IP "\f3Constructors\f1"
These are the same as those for \f(CWElement_Node\f1.
.IP "\f3Destructor\f1"
In addition to destroying itself,
the destructor also destroys all of its child nodes, if any.
.IP "\f(CWvoid add_child( HTML_Node *child )\fP"
If \f(CWchild\f1 is null
or it equals \f(CWthis\f1, does nothing.
Otherwise,
if the child node already has a parent,
it is removed from that parent's list of child nodes first;
then it is added to this node's list of child nodes
and sets the child's parent to \f(CWthis\f1.
.IP "\f(CWchild_list& children()\fP"
.IP "\f(CWchild_list const& children() const\fP"
Returns a reference to this node's list of child nodes,
or a reference to an empty list if there are none.
.IP "\f(CWbool empty() const\fP"
Returns true only if this node has no child nodes.
.IP "\f(CWbool remove_child( HTML_Node *child )\fP"
If \f(CWchild\f1 is null, does nothing and returns \f(CWfalse\f1.
Otherwise,
searches through this node's list of child nodes
looking for the child.
If found,
removes the child, set the child's parent to null, and returns \f(CWtrue\f1;
otherwise, returns \f(CWfalse\f1.
.IP "\f(CWvirtual void visit( visitor const &v, int depth = 0 )\fP"
This member function overrides \f(CWHTML_Node\f1's \f(CWvisit()\f1.
It simply calls \f(CWvisit()\f1 for each child node in order
passing \f(CWdepth\f1.
In pseudo-code:
.cS
	for ( \f2every child\fP )
		\f2child\fP->visit( v, depth );
.cE
It does
.I not
call \f(CWvisit()\f1 for itself.
It is therefore effectively ``invisible'' in the HTML node tree.
.SH SEE ALSO
.BR Element_Node (3),
.BR HTML_Node (3).
.SH AUTHOR
Paul J. Lucas
.RI < pjl@best.com >


Changes
HTML Tree Changes
=================

*******************************************************************************
1.2.3
*******************************************************************************

BUG FIXES
---------

* When setting an HTML attribute's value to a variable whose value was undef,
  the attribute was not deleted.

  (This bug fix shall be known as bug fix AVU.)


CHANGES, file-by-file
---------------------

* mod/HTML/Tree/Tree.xs

	1. In att(), performed the following substitution:

		s/new_value == &PL_sv_undef/!SvOK( new_value )/

	   for bug fix AVU.

* version.h

	1. Updated version to "1.2.3".


*******************************************************************************
1.2.2
*******************************************************************************

BUG FIXES
---------

* Fixed segmentation fault when parsing HTML files that contain tags longer
  than Tag_Name_Max_Size characters.

  (This bug fix shall be known as bug fix HTL.)


CHANGES, file-by-file
---------------------

* html.c

	1. In parse_html_tag(), added tag buffer overflow check for bug
	   fix HTL.

* version.h

	1. Updated version to "1.2.2".


*******************************************************************************
1.2.1
*******************************************************************************

BUG FIXES
---------

* Fixed a couple of bugs in mod/HTML/Tree/test.pl (see below).


CHANGES, file-by-file
---------------------

* mod/HTML/Tree/test.pl

	1. Simplified fetchrow() function and also eliminated dummy '0'
	   in data.

	2. Fixed path to test.html.

	3. Added "$out =~ s/^\s*\n//gm;" just before comparison to
	   accomodate differences in whitespace.

* version.h

	1. Changed version number to "1.2.1".


*******************************************************************************
1.2
*******************************************************************************

NEW FEATURES
------------

* Added a built-in function to substitute the HREF attribute as a shorthand.

  (This feature shall be known as feature SUBHREF.)

* Added a built-in function to substitute just the numeric ID part of an
  attribute's value.

  (This feature shall be known as feature SUBID.)

* I've decided that "sub::" makes more sense that "att::"

  (This change shall be known as change ATT2.)

* Changed the behavior when a CLASS attribute has multiple values for end tags
  in that only the first value's function is called.

  (This feature shall be known as feature CMVE.)


BUG FIXES
---------

* <DD> elements weren't implicitly terminated by a new <DT> element.

  (This bug fix shall be known as bug fix DDDT.)

* The "prettyhtml" test target didn't have its dependencies correctly in the
  GNUmakefile due to an error in the "include" directive.

  (This bug fix shall be known as bug fix IDT.)


CHANGES, file-by-file
---------------------

* elements.c

	1. Added:

		"dt", "/dt",

	   to the "dd" line for bug fix DDDT.

* GNUmakefile

	1. Performed following substitution:

		s/$(TEST:/$(TEST_SRCS:/

	   for bug fix IDT.

* mod/Apache/HTML/ClassParser/ClassParser.pm

	1. Performed following substitution:

		s/$VERSION = '1.0'/$VERSION = '1.2'/

	2. In sub_att(), performed following substitution:

		s/att::/sub::/

	   for change ATT2.

	3. Added sub_href function for feature SUBHREF and added to the
	   documentation for it.

	4. Added sub_att_id, sub_href_id, and sub_value_id functions
	   for feature SUBID and added to the documentation for it.

	5. In visitor(), changed it so that if the tag is an end tag,
	   return after calling the first CLASS value function (for
	   feature CMVE).

	6. Added description of for CLASS attributes that have multiple
	   values to the documentation.

	7. Added documention for the heretofore undocumented pm_uri feature.

	8. Fixed some grammar and spelling errors in the documentation.

* version.h

	1. Changed version number to "1.2".


*******************************************************************************
1.1.1
*******************************************************************************

BUG FIXES
---------

* "att::" was completely broken because it didn't do the right pattern match.
  I don't know how this ever worked.

  (This bug fix shall be known as bug fix ATT.)


CHANGES, file-by-file
---------------------

* mod/Apache/HTML/ClassParser/ClassParser.pm

	1. Simplifed append_att().

	2. In sub_att(), performed following substitution:

		s/sub::/att::/

	   for bug fix ATT.

* version.h

	1. Changed version number to "1.1.1".


*******************************************************************************
1.1
*******************************************************************************

CHANGES, file-by-file
---------------------

* config/config.mk

	1. Removed CCCEXT, et al.

	2. Moved manual page stuff to man.mk.

* config/man.mk

	1. Copied improved version from SWISH++ source tree.

* file_vector.h

	1. Replaced off_t with size_t due to size differences when
	   compiled under Perl 5.6 supporting large files.

* man/GNUakefile
* man/Makefile

	1. Renamed Makefile to GNUmakefile.

	2. Updated to use auto dependency generation.

* man/man3/GNUmakefile

	1. Added this missing file.

* mod/Apache/HTML/ClassParser/ClassParser.pm

	1. Changed pattern to look for a package such that it is found
	   only if the word package is precedded by zero or more
	   whitespace characters.

	2. For visitor(), added the missing comment for the RETURN
	   VALUE.

	3. In visitor(), made it so that "exists" is used with
	   function_map so new keys are not created.

	4. Minor tweaks, typo corrections, etc, in manual page.

* mod/HTML/Tree/Makefile.PL

	1. Performed following substitutions:

		s/-DHAS_BOOL/-Ubool/

	   to get rid of incorrect bool definition Perl introduces.

* mod/HTML/Tree/Tree.xs

	1. Simplified "if" statement in children().

	2. Performed following substitutions:

		s/na/PL_na/
		s/sv_undef/PL_sv_undef/


Parent_Node.c
/*
**	HTML Tree
**	Parent_Node.c
**
**	Copyright (C) 1999  Paul J. Lucas
**
**	This program is free software; you can redistribute it and/or modify
**	it under the terms of the GNU General Public License as published by
**	the Free Software Foundation; either version 2 of the License, or
**	(at your option) any later version.
** 
**	This program is distributed in the hope that it will be useful,
**	but WITHOUT ANY WARRANTY; without even the implied warranty of
**	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
**	GNU General Public License for more details.
** 
**	You should have received a copy of the GNU General Public License
**	along with this program; if not, write to the Free Software
**	Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

// local
#include "Parent_Node.h"

//*****************************************************************************
//
// SYNOPSIS
//
	/* virtual */ Parent_Node::~Parent_Node()
//
// DESCRIPTION
//
//	Destroy a Parent_Node and destroy all of our child nodes recursively.
//	Set our destructing_ flag so our children don't bother removing
//	themselves from our list of children since we're destroying ourselves
//	anyway.
//
//*****************************************************************************
{
	destructing_ = true;
	for ( child_list::iterator
		child = children_.begin(); child != children_.end(); ++child
	)
		delete *child;
}

//*****************************************************************************
//
// SYNOPSIS
//
	void Parent_Node::add_child( HTML_Node *new_child )
//
// DESCRIPTION
//
//	Add a given child node to our list of children taking care to remove it
//	from its current parent's list of children, if any.
//
// PARAMETERS
//
//	new_child	A pointer to the child node to be added.
//
//*****************************************************************************
{
	if ( new_child && new_child->parent_ != this ) {
		if ( new_child->parent_ )
			new_child->parent_->remove_child( new_child );
		new_child->parent_ = this;
		children_.push_back( new_child );
	}
}

//*****************************************************************************
//
// SYNOPSIS
//
	bool Parent_Node::remove_child( HTML_Node *orphan )
//
// DESCRIPTION
//
//	Remove a given child node from our list of children.
//
// PARAMETERS
//
//	orphan	A pointer to the child node to be removed.
//
// RETURN VALUE
//
//	Returns true only if the child was removed.
//
//*****************************************************************************
{
	if ( orphan )
		for ( child_list::iterator
			child  = children_.begin();
			child != children_.end(); ++child
		)
			if ( *child == orphan ) {
				children_.erase( child );
				orphan->parent_ = 0;
				return true;
			}
	return false;
}

//*****************************************************************************
//
// SYNOPSIS
//
	/* virtual */ void Parent_Node::visit( visitor const &v, int depth )
//
// DESCRIPTION
//
//	Have the visitor visit all of our children.  We intentionally do NOT
//	pass "depth + 1" because Parent_Nodes are supposed to be "invisible" in
//	the tree.
//	
// PARAMETERS
//
//	v	A reference to a visitor.
//
//*****************************************************************************
{
	for ( child_list::const_iterator
		child = children_.begin(); child != children_.end(); ++child
	)
		(*child)->visit( v, depth );
}
